File: demos/demo_leader.cpp
#include "synchronizer/leader.hpp"

#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <iostream>
#include <signal.h>


#define SYNC_MEM "sync_mem"
#define COUNT_MEM "count_mem"


// demo_follower or demo_follower.py should be started first.

// expected results:
// demo_follower runs at 3 Hz.
// when demo_leader starts, it switches to run to
// "as fast as possible" (start_sync is called)
// then back to 3Hz ("stop_sync" is called),
// then back to as fast as possible (start_sync)


void looping(synchronizer::Leader &s)
{
  for (int iteration=0;iteration<50000;iteration++)
    {
      int count;
      shared_memory::get<int>(COUNT_MEM,"count",count);
      count++;
      shared_memory::set<int>(COUNT_MEM,"count",count);
      std::cout << "\t\tleader: " << count << "\n";
      s.pulse();
    }
}


void execute()
{

  synchronizer::Leader s(SYNC_MEM);

  s.start_sync(1000000);
  looping(s);
  
  std::cout << "stopping sync mode for 2 seconds\n";
  s.stop_sync();
  usleep(2000000);
  std::cout << "restarting sync mode\n";

  s.start_sync(1000000);
  looping(s);
  
  std::cout << "EXIT\n";
  
  // turning off follower
  shared_memory::set<bool>(COUNT_MEM,"running",false);
  
}


int main(){

  execute();
  
}

File: demos/demo_leader.py
import synchronizer,time,shared_memory_py

leader = synchronizer.Leader("sync_mem")


# running in sync mode for 4 seconds.
# as fast as possible
print("sync: follower running as fast as possible")
leader.start_sync(-1)
time_start = time.time()
while time.time()-time_start < 4 :
    leader.pulse()

# stopping synchronization,
# follower can run at whatever frequency
# it is set to (in this case 3Hz)
leader.stop_sync()
print("async: follower running at dedicated frequency")
time.sleep(3)

# running in sync mode for 4 seconds.
# as fast as possible
print("sync: follower running as fast as possible")
leader.start_sync(-1)
time_start = time.time()
while time.time()-time_start < 4 :
    leader.pulse()

print("leader exit")
